File: LeastChange.js
const AMOUNTS = [
  {
    type: "one hundred dollar bill",
    value: 10000,
  },
  {
    type: "fifty dollar bill",
    value: 5000,
  },
  {
    type: "twenty dollar bill",
    value: 2000,
  },
  {
    type: "ten dollar bill",
    value: 1000,
  },
  {
    type: "five dollar bill",
    value: 500,
  },
  {
    type: "toonie",
    value: 200,
  },
  {
    type: "loonie",
    value: 100,
  },
  {
    type: "quarter",
    value: 25,
  },
  {
    type: "dime",
    value: 10,
  },
  {
    type: "nickel",
    value: 5,
  },
  {
    type: "penny",
    value: 1,
  },
]

function calculateChange(amount) {
  let placeInArray = 0
  let numberOfCurrentValue = 0
  let changeStrings = []
  while (amount > 0 && AMOUNTS.length > placeInArray) {
    if (amount < AMOUNTS[placeInArray].value) {
      placeInArray = placeInArray + 1
    } else {
      while (amount - AMOUNTS[placeInArray].value >= 0) {
        amount = amount - AMOUNTS[placeInArray].value
        numberOfCurrentValue = numberOfCurrentValue + 1
      }
      const string_returned = changeAsString(placeInArray, numberOfCurrentValue)
      numberOfCurrentValue = 0
      changeStrings.push(string_returned)
    }
  }

  outputChangeResult(changeStrings)
}

function outputChangeResult(changeStrings) {
  let changeStringsCount = 0
  let totalChangeRequired = ""
  if (changeStrings.length == 1) {
    console.log(changeStrings[0])
  } else {
    changeStrings.forEach(string => {
      changeStringsCount = changeStringsCount + 1
      if (changeStrings.length != changeStringsCount) {
        totalChangeRequired += string + ", "
      } else {
        totalChangeRequired += string
      }
    })
  }

  console.log(`\nChange required - ${totalChangeRequired}\n`)
}

function changeAsString(placeInArray, numberOfCurrentValue) {
  const currentString = numberOfCurrentValue + " " + AMOUNTS[placeInArray].type
  if (numberOfCurrentValue > 1) {
    return currentString + "s"
  }
  return currentString
}

if (!isNaN(Number(process.argv[2]))) {
  calculateChange(Number(process.argv[2]))
} else {
  console.log("Input must be an integer, the number of cents required in change.\n")
}

File: LeastChange.py
# GLOBALS

possible_values = [
  10000,
  5000,
  2000,
  1000,
  500,
  200,
  100,
  25,
  10,
  5,
  1
]

possible_string = [
  "one hundred dollar bill",
  "fifty dollar bill",
  "twenty dollar bill",
  "ten dollar bill",
  "five dollar bill",
  "toonie",
  "loonie",
  "quarter",
  "dime",
  "nickle",
  "penny"
]

def main():
  user_input = None
  first_visit = True
  while user_input != 'quit' and user_input != 'Q' and user_input != 'q':
    if first_visit:
      user_input = input("Welcome to Least Change, enter the $ amount in cents you require in change - ")
      first_visit = False
    else:
      user_input = input("Enter the $ amount in cents you require in change - ")

    input_amount = 0
    if user_input != 'quit' and user_input != 'Q' and user_input != 'q':
      try:
        input_amount = int(user_input)
      except ValueError:
        raise InvalidInput("Input must be an integer, the number of cents required in change.\n")

      calculate_change(input_amount)

  exit()

def calculate_change(amount):
  place_in_array = 0
  number_of_current_value = 0
  change_strings = []
  while amount > 0 and len(possible_values) > place_in_array:
    if amount < possible_values[place_in_array]:
      place_in_array = place_in_array + 1
    else:
      while amount - possible_values[place_in_array] >= 0:
        amount = amount - possible_values[place_in_array]
        number_of_current_value = number_of_current_value + 1
      string_returned = change_as_string(place_in_array, number_of_current_value)
      number_of_current_value = 0
      change_strings.append(string_returned)

  output_change_result(change_strings)

def output_change_result(change_strings):
  change_strings_count = 0
  total_change_required = ""
  if len(change_strings) == 1:
    print(change_strings[0])
  else:
    for string in change_strings:
      change_strings_count = change_strings_count + 1
      if len(change_strings) != change_strings_count:
        total_change_required = total_change_required + string + ", "
      else:
          total_change_required = total_change_required + string

    print("\nChange required - " + total_change_required + "\n")

def change_as_string(place_in_array, number_of_current_value):
  current_string = str(number_of_current_value) + " " + possible_string[place_in_array]
  if number_of_current_value > 1:
    return current_string + "s"
  return current_string


class InvalidInput(Exception):
  def __init__(self, msg):
    super(InvalidInput, self).__init__(msg)

main()
